File: buildSrc/src/main/java/com/example/AndroidProjectGradlePlugin.java
package com.example;

import com.android.build.gradle.AppExtension;

import org.gradle.api.JavaVersion;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile;

import java.util.ArrayList;

import kotlin.Unit;

public class AndroidProjectGradlePlugin implements Plugin<Project> {

    @Override
    public void apply(@NotNull Project target) {
        applyPlugins(target);
        configurePlugins(target);
        applyDependencies(target);
    }

    void applyPlugins(Project project) {
        project.getPluginManager().apply("kotlin-android");
        project.getPluginManager().apply("kotlin-kapt");
        project.getPluginManager().apply("kotlin-parcelize");
    }

    //    @SuppressWarnings("UnstableApiUsage")
    void configurePlugins(Project project) {
        AppExtension androidExtension = project.getExtensions().getByType(AppExtension.class);
        androidExtension.compileSdkVersion(Const.AndroidProject.compileSdkVersion);
        androidExtension.buildToolsVersion(Const.AndroidProject.buildToolsVersion);

        androidExtension.defaultConfig(defaultConfig -> {
            defaultConfig.minSdkVersion(Const.AndroidProject.minSdkVersion);
            defaultConfig.targetSdkVersion(Const.AndroidProject.targetSdkVersion);
            defaultConfig.versionCode(Const.AndroidProject.versionCode);
            defaultConfig.versionName(Const.AndroidProject.versionName);
            defaultConfig.testInstrumentationRunner(Const.AndroidProject.testInstrumentationRunner);
            defaultConfig.vectorDrawables(vectorDrawables -> {
                vectorDrawables.setUseSupportLibrary(true);
            });
        });

        androidExtension.compileOptions(compileOptions -> {
            compileOptions.setCoreLibraryDesugaringEnabled(true);
            compileOptions.setSourceCompatibility(JavaVersion.VERSION_1_8);
            compileOptions.setTargetCompatibility(JavaVersion.VERSION_1_8);
        });

        //noinspection UnstableApiUsage
        androidExtension.getBuildFeatures().setViewBinding(true);

        project.getTasks().withType(KotlinCompile.class, kotlinCompile -> {
            kotlinCompile.kotlinOptions(kotlinJvmOptions -> {
                kotlinJvmOptions.setLanguageVersion("1.4");
                kotlinJvmOptions.setJvmTarget(JavaVersion.VERSION_1_8.toString());

//                kotlinJvmOptions.setAllWarningsAsErrors(true);

                ArrayList<String> compilerArgs = new ArrayList<>(kotlinJvmOptions.getFreeCompilerArgs());
                // т.к. модули имеют одинаковые названия "api" и "impl", то говорим компилятору генерировать для
                // каждого модуля META-INF на основе группы+названия проекта, т.е. вместо "impl.kotlin_module" будет "modules.app.impl.kotlin_module"
                compilerArgs.add("-module-name");
                compilerArgs.add(project.getGroup() + "." + project.getName());
                // поддержка inline классов
                compilerArgs.add("-Xinline-classes");
                kotlinJvmOptions.setFreeCompilerArgs(compilerArgs);
                return Unit.INSTANCE;
            });
        });
    }

    void applyDependencies(Project project) {
    }
}

File: buildSrc/src/main/kotlin/com/example/AndroidProjectGradlePlugin.kt
package com.example

import com.android.build.gradle.BaseExtension
import com.example.settings.AndroidPluginSettings
import com.example.settings.DependenciesSettings
import org.gradle.api.JavaVersion
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.artifacts.dsl.DependencyHandler
import org.gradle.kotlin.dsl.configure
import org.gradle.kotlin.dsl.dependencies
import org.gradle.kotlin.dsl.embeddedKotlinVersion
import org.gradle.kotlin.dsl.withType
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

open class AndroidProjectGradlePlugin : AbstractProjectGradlePlugin() {
    override fun apply(target: Project) {
        applyPlugins(target)
        configurePlugins(target)
        applyDependencies(target)
    }

    override fun applyPlugins(project: Project) {
        super.applyPlugins(project)
        project.pluginManager.apply("kotlin-android")
        project.pluginManager.apply("kotlin-kapt")
        project.pluginManager.apply("kotlin-parcelize")
    }

    //    @SuppressWarnings("UnstableApiUsage")
    @Suppress("UnstableApiUsage")
    override fun configurePlugins(project: Project) {
        super.configurePlugins(project)
        // Конфигурация Android gradle plugin. Аналог блока android {} в groovy скрипте
        project.extensions.configure<BaseExtension> {
            compileSdkVersion(AndroidPluginSettings.compileSdkVersion)
            buildToolsVersion(AndroidPluginSettings.buildToolsVersion)
            defaultConfig {
                minSdkVersion(AndroidPluginSettings.minSdkVersion)
                targetSdkVersion(AndroidPluginSettings.targetSdkVersion)
                versionCode(AndroidPluginSettings.versionCode)
                versionName(AndroidPluginSettings.versionName)
                testInstrumentationRunner(AndroidPluginSettings.testInstrumentationRunner)
                vectorDrawables.useSupportLibrary = true
            }

            compileOptions {
                isCoreLibraryDesugaringEnabled = true
                setSourceCompatibility(JavaVersion.VERSION_1_8)
                setTargetCompatibility(JavaVersion.VERSION_1_8)
            }

            buildFeatures.viewBinding = true

            project.tasks.withType<KotlinCompile>().configureEach {
                kotlinOptions {
                    languageVersion = "1.4"
                    jvmTarget = "1.8"

//                allWarningsAsErrors = true

                    freeCompilerArgs = freeCompilerArgs + listOf(
                        // Т.к. модули имеют одинаковые названия ("api" и "impl"), то говорим компилятору генерировать
                        // для каждого модуля META-INF на основе группы+названия проекта,
                        // т.е. вместо "impl.kotlin_module" будет "modules.app.impl.kotlin_module".
                        // Поскольку мы делаем это в каждой таске модуля, то будем получать такой варнинг "Argument -module-name is passed multiple times. Only the last value will be used"
                        "-module-name", "${project.group}.${project.name}",
                        // поддержка inline классов
                        "-Xinline-classes"
                    )
                }
            }
        }
    }

    override fun applyDependencies(project: Project) {
        super.applyDependencies(project)
        project.dependencies {
            // desugaring
            addCoreLibraryDesugaring(DependenciesSettings.Android.desugaring)

            // kotlin
            addImplementation(
                DependenciesSettings.Jetbrains.kotlinStdlibJdk7(embeddedKotlinVersion),
                DependenciesSettings.Jetbrains.kotlinReflect(embeddedKotlinVersion)
            )

            // androidx
            addImplementation(
                DependenciesSettings.Androidx.legacySupport,
                DependenciesSettings.Androidx.appcompat,
                DependenciesSettings.Androidx.recyclerview,
                DependenciesSettings.Androidx.cardview,
                DependenciesSettings.Androidx.constraintlayout,
                DependenciesSettings.Androidx.material,
                DependenciesSettings.Androidx.viewpager2,
                DependenciesSettings.Androidx.coreKtx
            )

            // unit tests
            addTestImplementation(
                DependenciesSettings.Junit.junit,
                DependenciesSettings.Mockk.mockk
            )

            // dagger
            addCompileOnly(DependenciesSettings.Glassfish.javaxAnnotation)
            addKapt(DependenciesSettings.Dagger.daggerCompiler)
            addImplementation(DependenciesSettings.Dagger.dagger)

            // cicerone
            addImplementation(DependenciesSettings.Terrakok.cicerone)
        }
    }
}